BUG: rolling.quantile does not return an interpolated result

guillemborrell opened the pull request:
Now computing the quantile of a rolling window matches the default of the rolling method in Series, DataFrame and np.percentile.

Fixes bugs #9413 and #16211

 - [x] closes #9413
 - [x] tests added / passed
 - [x] passes ``git diff upstream/master --name-only -- '*.py' | flake8 --diff``
 - [x] whatsnew entry

guillemborrell committed:
BUG: rolling.quantile now returns an interpolated result (#9413)

jaimefrio commented on pandas/_libs/window.pyx:
```diff
@@ -1391,7 +1392,19 @@ def roll_quantile(ndarray[float64_t, cast=True] input, int64_t win,
+                qhigh = (<double> (idx + 1)) / (<double>(nobs - 1))
+                vlow = skiplist.get(idx)
+                vhigh = skiplist.get(idx + 1)
+                output[i] = vlow + (vhigh - vlow) * \
```
The computation time is probably dominated by the skip list, but it still feels wrong to have three divisions in the innermost loop. And I think we can actually get rid of all of them! If this is correct:

```
qlow = i / (nobs - 1)
qhigh = (i + 1) / (nobs - 1)
(quantile - qlow) / (qhigh - qlow) = (q - i / (nobs - 1)) / ((i + 1 - i) / (nobs - 1)) = (nobs - 1) * q - i
```

It should be possible to simply write

```
output[i] = vlow + (vhigh - vlow) * (quantile * (nobs - 1) - idx)
```

It should be faster, although probably not much, but on top of that I think that the fact that `quantile * (nobs - 1) - idx` is the fractional part of `quantile * (nobs - 1)` makes the code a little more self explanatory?

guillemborrell replied:
Now implemented as suggested.

jaimefrio commented on pandas/_libs/window.pyx:
```diff
@@ -1391,7 +1392,19 @@ def roll_quantile(ndarray[float64_t, cast=True] input, int64_t win,
-            output[i] = skiplist.get(idx)
+
+            # Exactly last point
+            if idx == nobs - 1:
```
This can only happen if `quantile == 1.0`, right? It might be worth removing the branch from the inner loop: I think we could actually redirect `quantile == 1.0` to `roll_max` and `quantile == 0.0` to `roll_min`? If I am not missing anything, these are asymptotically faster, O(n) vs O(n log w).

guillemborrell replied:
quantile == 1.0 and == 0.0 now fallback to roll_max and roll_min respectively

jaimefrio commented on pandas/_libs/window.pyx:
```diff
@@ -1391,7 +1392,19 @@ def roll_quantile(ndarray[float64_t, cast=True] input, int64_t win,
+
+            # Interpolated quantile
+            else:
+                qlow = (<double> idx) / (<double>(nobs - 1))
```
Bad things (TM) happen if `nobs == 1`, not only because of the obvious division by zero here, but `vhigh` will index the skiplist out of bounds. Unless `minp` can be guaranteed to be at `>= 2` upstream (I don't think it can!) we are going to have to guard against that possibility in the inner loop.

We could simply branch on `nobs >= 1`, which is probably more predictable and faster in the more common cases, or another possibility is to only compute `vhigh` and interpolate if `quantile * (nobs - 1) != idx`.

However this is fixed, a test with enough NaNs to have a single observation in some window, and `min_periods=1`,  should certainly be added.

guillemborrell replied:
The branch that handled quantile == 1.0 also handled nobs == 1, that now is handled separately in the innermost loop. There may be more optimal ways to handle this case in a more external loop, but IMHO this change is not that bad, since this condition only happens when there's a single loop.

I have not added more tests, since TestMoments::test_rolling_quantile tests the mentioned cases too.

jaimefrio reviewed (commented):
PTAL, I think there are some corner cases that need taking care off, and have also a couple of enhancement suggestions.

Aside from those it LGTM.

jreback commented on doc/source/whatsnew/v0.20.1.txt:
```diff
@@ -78,7 +78,7 @@ Reshaping
 ^^^^^^^
 
-
+- Bug in ``.rolling.quantile()`` which incorrectly used different defaults than Series.quantile() and DataFrame.quantile()
```
move to 0.21.0

guillemborrell replied:
Moved.

jreback reviewed (changes requested):
pls address @jaimefrio comments

Commit:
Implemented changes suggested by reviewer @jaimefrio

Commit:
Implemented changes suggested by reviewer @jaimefrio

guillemborrell reviewed (commented):
Please, check if all comments have been successfully addressed.

jaimefrio commented on pandas/_libs/window.pyx:
```diff
@@ -1348,6 +1348,7 @@ def roll_quantile(ndarray[float64_t, cast=True] input, int64_t win,
         ndarray[double_t] output
+        double vlow, vhigh
 
     if quantile < 0.0 or quantile > 1.0:
```
The code below now relies on 0.0 and 1.0 being filtered by the Python wrapper, doesn't it? We probably should make that explicit here by changing `<` and `>` to `<=` and `>=`.

guillemborrell replied:
Just wanted to note that this line was not altered by my patch

jaimefrio replied:
Yes, but your patch does alter the valid inputs, and `quantile == 1.0` is no longer a valid one. It's more of a cosmetic thing, because the Python wrapper never calls this with `quantile=1.0`, but I think it is still worth making the change.

jaimefrio commented on pandas/_libs/window.pyx:
```diff
@@ -1391,7 +1392,17 @@ def roll_quantile(ndarray[float64_t, cast=True] input, int64_t win,
+            else:
+                vlow = skiplist.get(idx)
+                vhigh = skiplist.get(idx + 1)
+                output[i] = vlow + (vhigh - vlow) * \
```
Very minor nitpick: there seems to be no other occurrence of `\` anywhere in this file. Can someone more familiar with Pandas style requirements comment on whether this is OK, or implicit line continuation inside parentheses is preferred?

jreback replied:
that's fine, or you can use (). and break it at an operator.

jaimefrio commented on pandas/_libs/window.pyx:
```diff
@@ -1391,7 +1392,17 @@ def roll_quantile(ndarray[float64_t, cast=True] input, int64_t win,
+            # Interpolated quantile
+            else:
+                vlow = skiplist.get(idx)
+                vhigh = skiplist.get(idx + 1)
```
Hadn't realized this before: `skiplist.get` is an O(log win) operation which we now have to perform twice. We also do a `skiplist.insert` and a `skiplist.remove`, which are also O(log win), and probably more expensive than `skiplist.get`, but a pessimistic estimation says the performance of this function will go down by 25%.

I think there are two things to do here:

 1. quantify that slow down by running a couple of benchmarks with and without this PR, and leave a record of the results for further reference.
 2. if the slowdown is significant, we could gain the speed back by doing some changes to the skiplist internals: when the first index is fetched, getting the next one is trivially simple and cheap, so we could add a `.get_two()` method to `IndexableSkiplist` and use it here. I think that would be work for another PR, but creating an issue for tracking based on the results of the benchmarks should be part of merging this.

guillemborrell replied:
I ran some very simple performance tests. The new version is indeed 25% slower.

*New*

```python
In [1]: import pandas
In [2]: pandas.__version__
Out[2]: '0.20.0+19.g36db9bdaf'
In [3]: import numpy as np
In [4]: df = pandas.DataFrame({'a': np.random.random(1000000)})
In [5]: %timeit df.a.rolling(10).quantile(0.5)
1 loop, best of 3: 1.78 s per loop
In [6]: %timeit df.a.rolling(10).median()
1 loop, best of 3: 460 ms per loop
```
*Old*

```python
In [1]: import pandas
In [2]: pandas.__version__
Out[2]: '0.20.1'
In [3]: import numpy as np
In [4]: df = pandas.DataFrame({'a': np.random.random(1000000)})
In [5]: %timeit df.a.rolling(10).quantile(0.5)
1 loop, best of 3: 1.4 s per loop
In [6]: %timeit df.a.rolling(10).median()
1 loop, best of 3: 451 ms per loop
```
YMMV, of course.

jaimefrio replied:
Thanks for doing this! Seeing this, there is a more obvious approach to speeding this in a future PR: use the C implemented skiplist that median uses, instead of the Cython `IndexableSkiplist` used by quantile. But that's clearly a different PR.

jreback replied:
@jaimefrio the issue is that for median you are only doing a single calculation. when doing *many* calculations (like here) a skip list is far faster. The actual impl of the skiplist could be improved a lot though (it has python objects inside it).

jaimefrio reviewed (approved):
I think this is very close to being ready, just a couple of nitpicks, and a suggestion for future improvements, but should be mostly ready to go.

guillemborrell committed:
quantile 1.0 and 0.0 are handled by the wrapper.

guillemborrell committed:
quantile 1.0 and 0.0 are handled by the wrapper.

guillemborrell reviewed (commented):
I think that this last commit successfully addresses all reviews.

jaimefrio commented:
It think it is fully ready to go now. Thanks Guillem!

jreback commented:
do we have an asv for this (or for window functions generally)?

would be straightforward to add if not

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -107,6 +107,7 @@ Reshaping
 Numeric
 ^^^^^^^
 
+- Bug in ``.rolling.quantile()`` which incorrectly used different defaults than Series.quantile() and DataFrame.quantile()
```
move to rolling section

guillemborrell replied:
Moved

guillemborrell committed:
Move entry in Whatsnew to Groupby/Resample/Rolling

guillemborrell committed:
Move entry in Whatsnew to Groupby/Resample/Rolling

jreback commented:
@guillemborrell actually would be great to add them here. just add a new file (copy existing one) and edit.

guillemborrell commented:
I'll write some benchmarks then, but I won't be able to do it until next weekend. When does v 0.21 freeze?

jreback commented:
@guillemborrell just released 0.20.1, so 0.21 will be a few months.

Commit:
Added benchmarks for rolling windows

Commit:
Added benchmarks for rolling windows

guillemborrell commented:
@jreback I have added a benchmark file that tests the methods of rolling windows. I have added small and large windows for all tests despite they are only relevant for quantiles. Please, take a look if those benchmarks are good enough.

jreback commented:
can you show the benchmark results?

guillemborrell commented:
Sure. This is for pandas 0.20.1

```
· Discovering benchmarks
· Running 48 total benchmarks (1 commits * 1 environments * 48 benchmarks)
[  0.00%] ·· Building for existing-py_usr_bin_python
[  0.00%] ·· Benchmarking existing-py_usr_bin_python
[  2.08%] ··· Running rolling.DataframeRolling.time_rolling_corr                    2.23s
[  4.17%] ··· Running rolling.DataframeRolling.time_rolling_count                  4.43ms
[  6.25%] ··· Running rolling.DataframeRolling.time_rolling_count_l                4.35ms
[  8.33%] ··· Running rolling.DataframeRolling.time_rolling_cov                     2.29s
[ 10.42%] ··· Running rolling.DataframeRolling.time_rolling_kurt                   3.63ms
[ 12.50%] ··· Running rolling.DataframeRolling.time_rolling_kurt_l                 4.34ms
[ 14.58%] ··· Running rolling.DataframeRolling.time_rolling_max                    3.59ms
[ 16.67%] ··· Running rolling.DataframeRolling.time_rolling_max_l                  3.81ms
[ 18.75%] ··· Running rolling.DataframeRolling.time_rolling_median                 2.73ms
[ 20.83%] ··· Running rolling.DataframeRolling.time_rolling_median_l               2.66ms
[ 22.92%] ··· Running rolling.DataframeRolling.time_rolling_min                    3.55ms
[ 25.00%] ··· Running rolling.DataframeRolling.time_rolling_min_l                  3.84ms
[ 27.08%] ··· Running rolling.DataframeRolling.time_rolling_quantile_0           110.07ms
[ 29.17%] ··· Running rolling.DataframeRolling.time_rolling_quantile_0_l         234.01ms
[ 31.25%] ··· Running rolling.DataframeRolling.time_rolling_quantile_1           129.11ms
[ 33.33%] ··· Running rolling.DataframeRolling.time_rolling_quantile_1_l         290.30ms
[ 35.42%] ··· Running rolling.DataframeRolling.time_rolling_quantile_median      123.17ms
[ 37.50%] ··· Running rolling.DataframeRolling.time_rolling_quantile_median_l    262.23ms
[ 39.58%] ··· Running rolling.DataframeRolling.time_rolling_skew                   5.45ms
[ 41.67%] ··· Running rolling.DataframeRolling.time_rolling_skew_l                 3.93ms
[ 43.75%] ··· Running rolling.DataframeRolling.time_rolling_std                    4.63ms
[ 45.83%] ··· Running rolling.DataframeRolling.time_rolling_std_l                  3.08ms
[ 47.92%] ··· Running rolling.DataframeRolling.time_rolling_sum                    2.87ms
[ 50.00%] ··· Running rolling.DataframeRolling.time_rolling_sum_l                  2.59ms
[ 52.08%] ··· Running rolling.SeriesRolling.time_rolling_corr                      6.35ms
[ 54.17%] ··· Running rolling.SeriesRolling.time_rolling_count                     7.68ms
[ 56.25%] ··· Running rolling.SeriesRolling.time_rolling_count_l                   7.37ms
[ 58.33%] ··· Running rolling.SeriesRolling.time_rolling_cov                       4.78ms
[ 60.42%] ··· Running rolling.SeriesRolling.time_rolling_kurt                      5.04ms
[ 62.50%] ··· Running rolling.SeriesRolling.time_rolling_kurt_l                    4.87ms
[ 64.58%] ··· Running rolling.SeriesRolling.time_rolling_max                       4.76ms
[ 66.67%] ··· Running rolling.SeriesRolling.time_rolling_max_l                     4.85ms
[ 68.75%] ··· Running rolling.SeriesRolling.time_rolling_median                    3.89ms
[ 70.83%] ··· Running rolling.SeriesRolling.time_rolling_median_l                  3.87ms
[ 72.92%] ··· Running rolling.SeriesRolling.time_rolling_min                       4.67ms
[ 75.00%] ··· Running rolling.SeriesRolling.time_rolling_min_l                     4.89ms
[ 77.08%] ··· Running rolling.SeriesRolling.time_rolling_quantile_0              114.80ms
[ 79.17%] ··· Running rolling.SeriesRolling.time_rolling_quantile_0_l            223.83ms
[ 81.25%] ··· Running rolling.SeriesRolling.time_rolling_quantile_1              140.83ms
[ 83.33%] ··· Running rolling.SeriesRolling.time_rolling_quantile_1_l            274.94ms
[ 85.42%] ··· Running rolling.SeriesRolling.time_rolling_quantile_median         125.02ms
[ 87.50%] ··· Running rolling.SeriesRolling.time_rolling_quantile_median_l       263.54ms
[ 89.58%] ··· Running rolling.SeriesRolling.time_rolling_skew                      4.77ms
[ 91.67%] ··· Running rolling.SeriesRolling.time_rolling_skew_l                    4.59ms
[ 93.75%] ··· Running rolling.SeriesRolling.time_rolling_std                       4.54ms
[ 95.83%] ··· Running rolling.SeriesRolling.time_rolling_std_l                     9.10ms
[ 97.92%] ··· Running rolling.SeriesRolling.time_rolling_sum                       3.74ms
[100.00%] ··· Running rolling.SeriesRolling.time_rolling_sum_l                     3.79ms
```

And the benchmark from this PR. 

```
· Discovering benchmarks
· Running 48 total benchmarks (1 commits * 1 environments * 48 benchmarks)
[  0.00%] ·· Building for existing-py_usr_bin_python
[  0.00%] ·· Benchmarking existing-py_usr_bin_python
[  2.08%] ··· Running rolling.DataframeRolling.time_rolling_corr                    2.22s
[  4.17%] ··· Running rolling.DataframeRolling.time_rolling_count                  4.28ms
[  6.25%] ··· Running rolling.DataframeRolling.time_rolling_count_l                4.31ms
[  8.33%] ··· Running rolling.DataframeRolling.time_rolling_cov                     2.24s
[ 10.42%] ··· Running rolling.DataframeRolling.time_rolling_kurt                   3.53ms
[ 12.50%] ··· Running rolling.DataframeRolling.time_rolling_kurt_l                 3.48ms
[ 14.58%] ··· Running rolling.DataframeRolling.time_rolling_max                    3.70ms
[ 16.67%] ··· Running rolling.DataframeRolling.time_rolling_max_l                  3.65ms
[ 18.75%] ··· Running rolling.DataframeRolling.time_rolling_median                 2.59ms
[ 20.83%] ··· Running rolling.DataframeRolling.time_rolling_median_l               2.58ms
[ 22.92%] ··· Running rolling.DataframeRolling.time_rolling_min                    3.52ms
[ 25.00%] ··· Running rolling.DataframeRolling.time_rolling_min_l                  3.67ms
[ 27.08%] ··· Running rolling.DataframeRolling.time_rolling_quantile_0           114.64ms
[ 29.17%] ··· Running rolling.DataframeRolling.time_rolling_quantile_0_l         215.85ms
[ 31.25%] ··· Running rolling.DataframeRolling.time_rolling_quantile_1           139.79ms
[ 33.33%] ··· Running rolling.DataframeRolling.time_rolling_quantile_1_l         337.24ms
[ 35.42%] ··· Running rolling.DataframeRolling.time_rolling_quantile_median      157.48ms
[ 37.50%] ··· Running rolling.DataframeRolling.time_rolling_quantile_median_l    268.86ms
[ 39.58%] ··· Running rolling.DataframeRolling.time_rolling_skew                   3.50ms
[ 41.67%] ··· Running rolling.DataframeRolling.time_rolling_skew_l                 3.68ms
[ 43.75%] ··· Running rolling.DataframeRolling.time_rolling_std                    3.37ms
[ 45.83%] ··· Running rolling.DataframeRolling.time_rolling_std_l                  5.29ms
[ 47.92%] ··· Running rolling.DataframeRolling.time_rolling_sum                    2.65ms
[ 50.00%] ··· Running rolling.DataframeRolling.time_rolling_sum_l                  2.55ms
[ 52.08%] ··· Running rolling.SeriesRolling.time_rolling_corr                      6.54ms
[ 54.17%] ··· Running rolling.SeriesRolling.time_rolling_count                     8.18ms
[ 56.25%] ··· Running rolling.SeriesRolling.time_rolling_count_l                   7.78ms
[ 58.33%] ··· Running rolling.SeriesRolling.time_rolling_cov                       4.79ms
[ 60.42%] ··· Running rolling.SeriesRolling.time_rolling_kurt                      6.16ms
[ 62.50%] ··· Running rolling.SeriesRolling.time_rolling_kurt_l                    5.02ms
[ 64.58%] ··· Running rolling.SeriesRolling.time_rolling_max                       9.15ms
[ 66.67%] ··· Running rolling.SeriesRolling.time_rolling_max_l                     5.31ms
[ 68.75%] ··· Running rolling.SeriesRolling.time_rolling_median                    4.27ms
[ 70.83%] ··· Running rolling.SeriesRolling.time_rolling_median_l                  4.30ms
[ 72.92%] ··· Running rolling.SeriesRolling.time_rolling_min                       6.82ms
[ 75.00%] ··· Running rolling.SeriesRolling.time_rolling_min_l                     5.69ms
[ 77.08%] ··· Running rolling.SeriesRolling.time_rolling_quantile_0              128.50ms
[ 79.17%] ··· Running rolling.SeriesRolling.time_rolling_quantile_0_l            252.35ms
[ 81.25%] ··· Running rolling.SeriesRolling.time_rolling_quantile_1              134.18ms
[ 83.33%] ··· Running rolling.SeriesRolling.time_rolling_quantile_1_l            280.08ms
[ 85.42%] ··· Running rolling.SeriesRolling.time_rolling_quantile_median         118.44ms
[ 87.50%] ··· Running rolling.SeriesRolling.time_rolling_quantile_median_l       276.01ms
[ 89.58%] ··· Running rolling.SeriesRolling.time_rolling_skew                      5.27ms
[ 91.67%] ··· Running rolling.SeriesRolling.time_rolling_skew_l                    5.01ms
[ 93.75%] ··· Running rolling.SeriesRolling.time_rolling_std                       7.04ms
[ 95.83%] ··· Running rolling.SeriesRolling.time_rolling_std_l                     5.39ms
[ 97.92%] ··· Running rolling.SeriesRolling.time_rolling_sum                       4.39ms
[100.00%] ··· Running rolling.SeriesRolling.time_rolling_sum_l                     4.23ms
```

jreback commented:
did the benchmarks show a diff? also pls rebase and push again just to confirm tests still working.

guillemborrell commented:
No significant changes other than rolling windows for quantiles, that are about 25% slower, but YMMV.

jreback commented:
looks like you picked up some other changes, do

```
git rebase -i upstream/master
git push <yourbranchname> <yourfork> -r
```

jreback commented on doc/source/whatsnew/v0.21.0.txt:
```diff
@@ -167,9 +167,11 @@ Plotting
 
+- Bug in ``DataFrame.resample().size()`` where an empty ``DataFrame`` did not return a ``Series`` (:issue:`14962`)
 - Bug in ``infer_freq`` causing indices with 2-day gaps during the working week to be wrongly inferred as business daily (:issue:`16624`)
+- Bug in ``.rolling.quantile()`` which incorrectly used different defaults than Series.quantile() and DataFrame.quantile()
```
``.rolling(...).quantile()`` and use :func:`Series.quantile` and :func:`DataFrame.quantile`

jreback replied:
add the issue number #9413 and #16211

jreback commented on pandas/tests/test_window.py:
```diff
@@ -1138,6 +1149,28 @@ def alt(x):
             self._check_moment_func(f, alt, name='quantile', quantile=q)
 
+    def test_rolling_quantile_np_percentile(self):
+        # #9413
```
add a 1-liner to both functions explaining what you are testing (e.g. the nature of the bug)

jreback reviewed (changes requested):
doc comments. ping on green.

guillemborrell committed:
Fixed whatsnew syntax and comments on new tests.

guillemborrell committed:
Fixed trailing whitespace to make the linter happy

guillemborrell commented:
It seems that app appveyor failed, but not due to a test. How do I proceed?

jreback commented:
i restarted

jaimefrio commented:
It's all green now.

guillemborrell commented:
Thanks for pinging @jaimefrio. I was off-screen for a couple of days.

jreback commented:
thanks!
